proposals/NNNN-inheritance-of-actor-isolation.md

```diff
@@ -0,0 +1,551 @@
+context, which can be a serious expressivity restriction, especially
+for higher-order functions.  It may also cause unwanted suspensions.
+
+[^2]: Prior to SE-0338, non-isolated asynchronous functions still
+didn't properly inherit their caller's isolation: they just didn't
+actively switch away.  As a result, they ran with whatever isolation
+they happened to the called or resumed with.  That is not good enough
+to allow them to safely be passed actor-isolated data or to make
+strong guarantees of a lack of suspensions.  This is now an ABI
+constraint: even if we wanted to change the language to make these
+functions inherit their caller's isolation by default, they aren't
+passed that information reliably and have no way to implement those
+semantics.
+
+For example, suppose you wanted to write a version of `map` that
+allows the transform to be `async`:[^3]
+
+```swift
+extension Collection {
+  func sequentialMap<R>(transform: (Element) async -> R) async -> [R] {
+    var results: [R] = []
+    for elt in self {
+      results.append(await transform(elt))
+    }
+    return results
+  }
+}
+```
+
+[^3]: The prospect of an `async` `map` function naturally raises the
+question of whether it should be a *parallel* map, i.e. whether the
+operation should be run concurrently for different elements.
+Parallel maps are a useful but different operation from the one
+we're discussing here, which still runs the transform sequentially
+for each element.  Inheriting isolation isn't as important for a
+parallel map because the transform probably isn't isolated itself
+and certainly has to be `@Sendable`.
+
+As written above, `sequentialMap` is a non-isolated asynchronous
+function.  If it is called from another non-isolated asynchronous
+function, everything's okay: it can be passed an arbitrary function
+and work with arbitrary types.  But if it's called from an *isolated*
+asynchronous function, Swift will treat the call as crossing an
+isolation barrier and enforce three restrictions:
+
+- First, the result of the call (`[R]`, i.e. `Array<R>`) must be
+  `Sendable`.  `Array` conditionally conforms to `Sendable` if its
+  `Element` type does, so this means `R` must be `Sendable`.
+  In short, this restriction prevents `sequentialMap` from being
+  used from an actor to produce an array of values isolated to the
+  actor.
+
+- Second, the `self` argument to the call (the collection) must
+  be `Sendable`.  Collections are generally `Sendable` if their
+  element types are `Sendable`, so in short, this restriction
+  prevents `sequentialMap` from being used from an actor to work
+  on a collection of values isolated to the actor.
+
+- Finally, the `transform` argument to the call must be `Sendable`.
+  Functions are generally `Sendable` if they don't capture any
+  `Sendable` values, so in short, this restriction prevents
+  `sequentialMap` from using any other data that's isolated to
+  the actor.
```

Comment:
I think there's a typo here, I'm not 100% sure about this but I'm pretty sure that if a function captured only sendable values (or capture nothing) the function would be sendable as well. So shouldn't this say this instead?
```suggestion
- Finally, the `transform` argument to the call must be `Sendable`.
  Functions are generally `Sendable` if they don't capture any
  non-`Sendable` values, so in short, this restriction prevents
  `sequentialMap` from using any other data that's isolated to
  the actor.
```

Reply:
You're right, thank you! I've reworded this, is it clearer now?